### PR TITLE
Carrie Walsh

### PR DESCRIPTION
**Pull request by carriewalsh**
Week 4 Indep. Assessment

### CODE REVIEW

File: lib/book.rb [hunk 1]
```diff
@@ -1,4 +1,14 @@
 class Book
+  attr_reader :author_first_name,
+      :author_last_name,
+      :title,
+      :publication_date
+  attr_writer :title
```
Comment (BrianZanti) [inline lib/book.rb]: I don't think this needs to be an attr_writer

File: lib/author.rb [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+      :last_name,
+      :books
 
+  def initialize(info)
+    @first_name = info[:first_name]
+    @last_name = info[:last_name]
+    @books = []
+  end
+
+  def add_book(title,publication_date)
```
Comment (BrianZanti) [inline lib/author.rb]: This method looks good! The only thing I would suggest is to put each of the attributes in the Hash on its own line for readability.

File: lib/library.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    @authors = Hash.new {|hash,key| hash[key] = []}
+  end
+
+  def add_to_collection(book)
+    @books << book
+    @authors["#{book.author_first_name} #{book.author_last_name}"] << book
+  end
+
+  def include?(title)
+    @books.any? do |book|
```
Comment (BrianZanti) [inline lib/library.rb]: 👍

File: lib/library.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      book.title == title
+    end
+  end
+
+  def card_catalog
+    no_the = @books.map do |book|
+      book.title = book.title.split
```
Comment (BrianZanti) [inline lib/library.rb]: It's good that you're thinking about how to ignore "The" in the title as an edge case, but I wouldn't go about it this way because you are permanently changing the book's title.

File: lib/library.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      if book.title[0] == "The"
```
Comment (BrianZanti) [inline lib/library.rb]: If the books title doesn't start with "The", it remains an array of strings. You should put the `join(" ")` after the if so that all the titles get joined together.

File: lib/library.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    end
+  end
+
+  def find_by_author(author)
+    books = []
+    if @authors.include?(author)
+      books = @authors[author]
+    end
+    books.flatten!
+    hash = books.group_by do |book|
```
Comment (BrianZanti) [inline lib/library.rb]: This isn't going to give you exactly what you want because group_by's values are arrays. A lot of times, it is best to avoid using an enumerable unless it does *exactly* what you want. `group_by` falls into that category pretty often.

File: test/author_test.rb [hunk 7]
```diff
@@ -1,6 +1,36 @@
...
+
+
+  def test_author_starts_with_no_books_by_default
+    author = Author.new({first_name: "Namey", last_name: "Name"})
+    assert_equal [], author.books
+  end
+
+
+  def test_author_can_have_books_added
+    assert_equal [@fifth_season, @kingdoms], @nk_jemisin.books
```
Comment (BrianZanti) [inline test/author_test.rb]: This test is a bit confusing because the method you are testing is actually called in your setup. Instead, you should call that method inside this test.

File: test/book_test.rb [hunk 8]
```diff
@@ -1,6 +1,33 @@
...
+  end
+
+
+  def test_book_has_title
+    assert_equal "To Kill a Mockingbird", @book.title
+  end
+
+
+  def test_book_has_publication_date
+    assert_equal "July 11, 1960", @book.publication_date
```
Comment (BrianZanti) [inline test/book_test.rb]: You need to convert the publication date to just the year.

File: test/library_test.rb [hunk 9]
```diff
@@ -1,6 +1,75 @@
...
+
+
+  def test_library_starts_with_no_books
+    lib = Library.new
+    assert_equal [], lib.books
+  end
+
+
+  def test_library_can_have_books_added
+    assert_equal [@fifth_season, @mockingbird, @kingdoms], @dpl.books
```
Comment (BrianZanti) [inline test/library_test.rb]: Similar problem to the AuthorTest here. Your Setup should only create objects and call methods that **aren't** defined on the class you're testing. Any methods you need to call on the class you're testing should be inside your tests.

File: test/library_test.rb [hunk 10]
```diff
@@ -1,6 +1,75 @@
...
+    assert_equal false, @dpl.include?("Grapes of Wrath")
+  end
+
+
+  def test_library_has_catalog_by_authors_last_name
+    assert_equal [@fifth_season,@kingdoms,@mockingbird], @dpl.card_catalog
+  end
+
+
+  def test_library_card_catalog_ignores_first_word_the
```
Comment (BrianZanti) [inline test/library_test.rb]: I like that you're thinking about this edge case, but I would prefer that you finish the challenge based on the spec, then go back and implement edge cases. You can go down too many rabbit holes if you try to think of *all* the edge cases.

**Comment by BrianZanti** [comment]
Great work Carrie!
